src/Assignments/AttachmentsSidebar.tsx

```diff
@@ -0,0 +1,178 @@
+    const [attachedFiles, setAttachedFiles] = useState<Array<{file: File, progress: number}>>([]);
+
+    // TODO: Get list of attached files.
+    useEffect(()=>{
+        const res: any[] = [];
+        setAttachedFiles(res.map(file => ({file: file, progress: 100})));
+    }, [topic]);
+
+    const updateIndexProgressWithOffset = (index: number, value: number, offset: number) => {
+        setAttachedFiles(attachedFiles => {
+            if (index >= attachedFiles.length) {
+                return attachedFiles;
+            }
+
+            const localAttachedFiles = [...attachedFiles];
+            localAttachedFiles[index] = {...attachedFiles[index], progress: offset + value};
+            return localAttachedFiles;
+        });
+    };
+
+    const updateIndexProgress = (index: number, progressEvent: any) => {
+        setAttachedFiles(attachedFiles => {
+            if (index >= attachedFiles.length) {
+                logger.error('Attempted to update progress beyond array bounds. (TSNH)', index);
+                return attachedFiles;
+            }
+            const progressPercent = Math.round((progressEvent.loaded / progressEvent.total) * 70);
+            const localAttachedFiles = [...attachedFiles];
+            localAttachedFiles[index] = {...attachedFiles[index], progress: 10 + progressPercent};
+            return localAttachedFiles;
+        });
+    };
+
+    const uploadFilesWIthProgress = async (attachedFiles: Array<{file: File, progress: number}>) => {
+        attachedFiles.forEach(async (file, index) => {
+            // Skip in-progress files.
+            if (file.progress > 0) return;
+            try {
+                const onUploadProgress = _.partial(updateIndexProgress, index);
+                const res = await getUploadURL();
+                logger.debug(res);
+                updateIndexProgressWithOffset(index, 20, 0);
+                    
+                await putUploadWork({
+                    presignedUrl: res.data.data.uploadURL,
+                    file: file.file, 
+                    onUploadProgress: onUploadProgress
+                });
+
+                await postConfirmAttachmentUpload({
+                    attachment: {
+                        cloudFileName: res.data.data.cloudFilename,
+                        userLocalFilename: file.file.name,
+                    },
+                    ...(gradeInstanceId ?
+                        {studentGradeInstanceId: gradeInstanceId} :
+                        {studentGradeId: gradeId}
+                    ),
+                });
+
+                updateIndexProgressWithOffset(index, 10, 90);
+            } catch (e) {
+                // Catch on an individual file basis
+                updateIndexProgressWithOffset(index, -1, 0);
+                logger.warn('A user encountered an error during attachment upload.', e.message);
```

Comment:
Should this be a warning or an error? Is there a common case in which uploads should fail?

Reply:
I had put it was a warning because the user gets some validation for it, so they can know to try it again, but I guess there isn't a valid fail case besides network issues.